### DISCUSSION TITLE
Decorator self-override operations not allowed for user-defined classes?

### DISCUSSION THREAD

**Discussion by NCBM** [microsoft/pyright#9637] [Q&A] [answered]
I'm writing some code like this:

```python
from typing import Callable, Self, reveal_type

class A[T]:
    def do_nothing(self, b: T) -> Self:
        ...

    def replace[Tx](self, b: Tx) -> "A[Tx]":
        ...

a = A[Callable[[], None]]()  # here is an error but ignore it for now

@a.do_nothing
def a() -> None:  # redeclaration error, have to ignore it since it is intentional
    ...

reveal_type(a)  # Type of "a" is "A[() -> None]"

@a.replace
def a(foo: int) -> None:  # error because of this
    ...

reveal_type(a)  # Type of "a" is "A[(foo: int) -> None]" # type changed intentionally
```

[View in pyright-play.net](https://pyright-play.net/?strict=true&locale=en-us&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDCAhgDYlEBGJApgDRQDK1Jw9I1AbtaQPrwLUAUIIDG5AM7ioAQQDaAFQC6ALkFR1UACbVgWsDxRgYAC1RoAFOOasoFZVHkBKKAFoAfI2uqNPqADoA4R9tXXYEchFqBQAPRUtrejsHaOd3KAAiOXlY9O9fdQC-YSIoAF4ZWWIyShpZWUV6ADkwFGpFOOcoAGIoY2p2TCkiFCh%2B8BBbAFd8JDRDAaR8YFwoQwB3YQABIj9NfUMTM0EQqCJzVI9m1vtuqHZtMSIQIhgkFtGQcfpjIi5YMEwc1w1Ew%2BHEqEioMGmBQMGosLeKFIag0hWE7C4vH41HMRE6PXkiBBYF06SI6WhmVk51clxabXSWx2YQiQhOZ2WYHsqBgFygV2oNx6YxWFGoIiIkysUBJsFM4hRBUCggx3BIfCJuPxDiJMtJ5MpcnMnO5sL5AsUFJ62KgIh%2B6GomhhcIRLVIJDggiAA)

The similar pattern appears on `@property` with `xxx.setter` or `xxx.deleter` which reports no errors. I wonder if it is a special case, and are there any way without writing ignore comments to reach the same effect.

**Comment by erictraut** [comment, marked as answer]
The `property` class requires extensive special-casing within type checkers and is exempted from redeclaration checks. 

I recommend against using this pattern in your own classes if you want them to work well with static typing. If you choose to use this pattern, you'll need to handle the ramifications of the redeclarations. 

One workaround is to avoid using the same symbol name for different type declarations, like this:

```python
from typing import Callable, Self

class A[T]:
    def do_nothing(self, b: T) -> Self: ...
    def replace[Tx](self, b: Tx) -> "A[Tx]": ...

a1 = A[Callable[[], None]]()

@a1.do_nothing
def a2() -> None: ...

@a2.replace
def a(foo: int) -> None: ...
```
